net: lib: wifi_mgmt_ext (#9261)
nrfconnect/sdk-nrf: doc/nrf/libraries/networking/wifi_credentials.rst

.. _lib_wifi_credentials:

> WiFi credentials
> ################

Reviewer: ```suggestion
Wi-Fi credentials
#################
```

.. contents::
   :local:
   :depth: 2

> The WiFi credentials library provides means to load and store WiFi network credentials.

Reviewer: ```suggestion
The Wi-Fi credentials library provides means to load and store Wi-Fi network credentials.
```

Overview
********

This library uses either Zephyr's settings subsystem or PSA protected storage to store credentials.
It also holds a list of SSIDs in RAM to provide dictionary-like access with SSIDs as keys.

Configuration
*************

> To use the WiFi credentials library, enable the :kconfig:option:`CONFIG_WIFI_CREDENTIALS` Kconfig option.

Reviewer: ```suggestion
To use the Wi-Fi credentials library, enable the :kconfig:option:`CONFIG_WIFI_CREDENTIALS` Kconfig option.
```

> You can pick a backend using following options:

Reviewer: ```suggestion
You can pick the backend using the following options:
```

* :kconfig:option:`CONFIG_WIFI_CREDENTIALS_BACKEND_PSA`
* :kconfig:option:`CONFIG_WIFI_CREDENTIALS_BACKEND_SETTINGS`

> The maximum number of networks is configured using the :kconfig:option:`CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES` Kconfig option.

Reviewer: ```suggestion
To configure the maximum number of networks, use the :kconfig:option:`CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES` Kconfig option.
```

The 802.11 standard does not specify the maximum length of SAE passwords.

> You can change the default using the :kconfig:option:`CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH` Kconfig option.

Reviewer: ```suggestion
To change the default, use the :kconfig:option:`CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH` Kconfig option.
```

> When using the PSA protected storage backend, entries will be saved under consecutive UIDs with an offset that is configured using :kconfig:option:`CONFIG_WIFI_CREDENTIALS_BACKEND_PSA_UID_OFFSET`.

Reviewer: ```suggestion
When using the PSA protected storage backend, entries will be saved under consecutive UIDs with an offset configured with the :kconfig:option:`CONFIG_WIFI_CREDENTIALS_BACKEND_PSA_UID_OFFSET` Kconfig option.
```

Adding credentials
******************

> Credentials can be added using the :c:func:`wifi_credentials_set_personal` and :c:func:`wifi_credentials_set_personal_struct` functions. The difference is that the former will build the internally used struct from given fields, while the latter takes the struct directly.

Reviewer: ```suggestion
You can add credentials using the :c:func:`wifi_credentials_set_personal` and :c:func:`wifi_credentials_set_personal_struct` functions.
The former will build the internally used struct from given fields, while the latter takes the struct directly.
```

If you add credentials with the same SSID twice, the older entry will be overwritten.

Querying credentials
********************

> Given an SSID, you can query credentials using the :c:func:`wifi_credentials_get_by_ssid_personal` and :c:func:`wifi_credentials_get_by_ssid_personal_struct` functions.

Reviewer: ```suggestion
With an SSID, you can query credentials using the :c:func:`wifi_credentials_get_by_ssid_personal` and :c:func:`wifi_credentials_get_by_ssid_personal_struct` functions.
```

> You can iterate over all stored credentials with :c:func:`wifi_credentials_for_each_ssid`.

Reviewer: ```suggestion
You can iterate over all stored credentials with the :c:func:`wifi_credentials_for_each_ssid` function.
```

> It is allowed to delete or overwrite credentials while iterating, since these operations do not change internal indices.

Reviewer: ```suggestion
Deleting or overwriting credentials while iterating is allowed, since these operations do not change internal indices.
```

Removing credentials
********************

> Credentials can be removed with the :c:func:`wifi_credentials_delete_by_ssid` function.

Reviewer: ```suggestion
You can remove credentials using the :c:func:`wifi_credentials_delete_by_ssid` function.
```

Limitations
***********

> 

Reviewer: It is not a good practise to have just a bullet list after a heading without any introductory text.

> 

Author: done

.. _lib_wifi_credentials:

Wi-Fi credentials
#################

.. contents::
   :local:
   :depth: 2

The Wi-Fi credentials library provides means to load and store Wi-Fi network credentials.

Overview
********

This library uses either Zephyr's settings subsystem or Platform Security Architecture (PSA) protected storage to store credentials.

> It also holds a list of SSIDs in RAM to provide dictionary like access with SSIDs as keys.

Reviewer: ```suggestion
It also holds a list of SSIDs in RAM to provide dictionary-like access using SSIDs as keys.
```

Configuration
*************

To use the Wi-Fi credentials library, enable the :kconfig:option:`CONFIG_WIFI_CREDENTIALS` Kconfig option.

You can pick the backend using the following options:

* :kconfig:option:`CONFIG_WIFI_CREDENTIALS_BACKEND_PSA`
* :kconfig:option:`CONFIG_WIFI_CREDENTIALS_BACKEND_SETTINGS`

To configure the maximum number of networks, use the :kconfig:option:`CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES` Kconfig option.

> The IEEE802.11 standard does not specify the maximum length of SAE passwords.

Reviewer: ```suggestion
The IEEE 802.11 standard does not specify the maximum length of SAE passwords.
```

Adding credentials
******************

You can add credentials using the :c:func:`wifi_credentials_set_personal` and :c:func:`wifi_credentials_set_personal_struct` functions.
The former will build the internally used struct from given fields, while the latter takes the struct directly.
If you add credentials with the same SSID twice, the older entry will be overwritten.

Querying credentials
********************

With an SSID, you can query credentials using the :c:func:`wifi_credentials_get_by_ssid_personal` and :c:func:`wifi_credentials_get_by_ssid_personal_struct` functions.

You can iterate over all stored credentials with the :c:func:`wifi_credentials_for_each_ssid` function.
Deleting or overwriting credentials while iterating is allowed, since these operations do not change internal indices.

Removing credentials
********************

You can remove credentials using the :c:func:`wifi_credentials_delete_by_ssid` function.

Limitations
***********

The library has the following limitations:

> * Although permitted by the IEEE802.11 standard, this library does not support zero-length SSIDs.

Reviewer: ```suggestion
* Although permitted by the IEEE 802.11 standard, this library does not support zero-length SSIDs.
```

* Wi-Fi Protected Access (WPA) Enterprise credentials are not yet supported.

> * The number of networks stored is fixed compile-time.

Reviewer: ```suggestion
* The number of networks stored is fixed compile time.
```

API documentation
*****************

| Header file: :file:`include/net/wifi_credentials.h`
| Source files: :file:`subsys/net/lib/wifi_credentials`

.. doxygengroup:: wifi_credentials

>    :project: nrf

Reviewer: Make sure you have the doxygengroup fixed also in the header file.
